<a href="https://colab.research.google.com/github/JobayerMahin/ETE-456-Project-Reports/blob/main/1608026_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Experiment Name: Exploring Different Types of Regression**
Course No: ETE 456

Name: Jobayer Alam Mahin

 ID: 1608026

**Objective**

> 1.To apply different regression algorithms in real world dataset.

**Task**

> Prediction of Insurance Costs

In [1]:

import warnings
warnings.filterwarnings("ignore")

**Importing Libraries**

In [2]:
import numpy as np        
import pandas as pd     
import matplotlib.pyplot as plt 

**Dataset**

In [3]:
# Download the data
!wget -O insurance.csv https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0

--2021-12-20 06:06:12--  https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.6.18, 2620:100:6022:18::a27d:4212
Connecting to www.dropbox.com (www.dropbox.com)|162.125.6.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/mwgqgjbmfw0xa5p/insurance.csv [following]
--2021-12-20 06:06:12--  https://www.dropbox.com/s/raw/mwgqgjbmfw0xa5p/insurance.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uca83548d3e13627adebc2dd6fa4.dl.dropboxusercontent.com/cd/0/inline/BcKzyMlpqv5FW3YcxVSttppHYdQoJ93nV9nOe0gW6VBNL8_KH-2qJwoPlJztdleo-r9fJu9MrZ6L7yXj-U4A9No4l7PI9wsigRXAOkpgz_i2SsWI7HsXfTD5OQ4Ylu22Gh7W5ZbIPx7Ha0r8reNORduR/file# [following]
--2021-12-20 06:06:13--  https://uca83548d3e13627adebc2dd6fa4.dl.dropboxusercontent.com/cd/0/inline/BcKzyMlpqv5FW3YcxVSttppHYdQoJ93nV9nOe0gW6VBNL8_KH-2qJwoPlJztdleo-r9fJu9MrZ6L7yXj-

**Importing Dataset**

In [4]:
dataset = pd.read_csv('insurance.csv')
dataset

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

Here, dataset has to be splitted into independent and dependent variables. Independent are the features which are used for prediction. Dependent is the target which will be predicted.

In [5]:
features = dataset[['age'	,'sex',	'bmi','children','smoker','region']]  # for independent
target = dataset[['charges']]   # for dependent

In [6]:
features

age     sex     bmi  children smoker     region
0      19  female  27.900         0    yes  southwest
1      18    male  33.770         1     no  southeast
2      28    male  33.000         3     no  southeast
3      33    male  22.705         0     no  northwest
4      32    male  28.880         0     no  northwest
...   ...     ...     ...       ...    ...        ...
1333   50    male  30.970         3     no  northwest
1334   18  female  31.920         0     no  northeast
1335   18  female  36.850         0     no  southeast
1336   21  female  25.800         0     no  southwest
1337   61  female  29.070         0    yes  northwest

[1338 rows x 6 columns]

In [7]:
target

charges
0     16884.92400
1      1725.55230
2      4449.46200
3     21984.47061
4      3866.85520
...           ...
1333  10600.54830
1334   2205.98080
1335   1629.83350
1336   2007.94500
1337  29141.36030

[1338 rows x 1 columns]

**Taking Care of Missing Data**

    Here 'sklearn' means sckit learn which is a package used for machine learning model. It contains many libraries.

preprocessing is a library which contains several classes,methods. From this library we import the Imputer class which will allow us to take care of the missing data


In [8]:
from sklearn.impute import SimpleImputer

In [9]:
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean") # imputer is an object of Imputer class 
imputer = imputer.fit(features[['age','bmi','children']])

In [10]:
features[['age','bmi','children']]= imputer.transform(features[['age','bmi','children']])

In [11]:
features

age     sex     bmi  children smoker     region
0     19.0  female  27.900       0.0    yes  southwest
1     18.0    male  33.770       1.0     no  southeast
2     28.0    male  33.000       3.0     no  southeast
3     33.0    male  22.705       0.0     no  northwest
4     32.0    male  28.880       0.0     no  northwest
...    ...     ...     ...       ...    ...        ...
1333  50.0    male  30.970       3.0     no  northwest
1334  18.0  female  31.920       0.0     no  northeast
1335  18.0  female  36.850       0.0     no  southeast
1336  21.0  female  25.800       0.0     no  southwest
1337  61.0  female  29.070       0.0    yes  northwest

[1338 rows x 6 columns]

In [12]:
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean") # imputer is an object of Imputer class 
imputer = imputer.fit(target[['charges']])

In [13]:
target[['charges']]= imputer.transform(target[['charges']])

In [14]:
target

charges
0     16884.92400
1      1725.55230
2      4449.46200
3     21984.47061
4      3866.85520
...           ...
1333  10600.54830
1334   2205.98080
1335   1629.83350
1336   2007.94500
1337  29141.36030

[1338 rows x 1 columns]

**One Hot Encoding**

  One hot encoder is used to convert the categorical data into numerical, so that they can be used in prediction.



In [16]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder
encoder=OneHotEncoder(sparse=False)
encoded_labels = pd.DataFrame (encoder.fit_transform(features[['sex','smoker','region']]))
encoded_labels.columns = encoder.get_feature_names(['sex','smoker','region'])
dataset= pd.concat([features, encoded_labels ], axis=1)

In [17]:
dataset

age     sex  ...  region_southeast  region_southwest
0     19.0  female  ...               0.0               1.0
1     18.0    male  ...               1.0               0.0
2     28.0    male  ...               1.0               0.0
3     33.0    male  ...               0.0               0.0
4     32.0    male  ...               0.0               0.0
...    ...     ...  ...               ...               ...
1333  50.0    male  ...               0.0               0.0
1334  18.0  female  ...               0.0               0.0
1335  18.0  female  ...               1.0               0.0
1336  21.0  female  ...               0.0               1.0
1337  61.0  female  ...               0.0               0.0

[1338 rows x 14 columns]

In [18]:
dataset.columns

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'sex_female',
       'sex_male', 'smoker_no', 'smoker_yes', 'region_northeast',
       'region_northwest', 'region_southeast', 'region_southwest'],
      dtype='object')

In [19]:
new_features = dataset[['age', 'sex_female', 'sex_male',
       'smoker_no', 'smoker_yes', 'bmi','children','region_northeast','region_northwest','region_southeast','region_southwest']]

**Splitting the dataset**

In [26]:
from sklearn.model_selection import train_test_split

X_train,X_test,y_train,y_test=train_test_split(new_features,target,test_size=0.15,random_state=0)
# random_state = 0 is select to get the same result

In [27]:
print(X_train.shape)
print(X_test.shape)

(1137, 11)
(201, 11)


In [28]:
X_train

age  sex_female  ...  region_southeast  region_southwest
420   64.0         0.0  ...               1.0               0.0
27    55.0         1.0  ...               0.0               0.0
982   31.0         0.0  ...               0.0               1.0
1005  29.0         0.0  ...               0.0               0.0
615   47.0         1.0  ...               1.0               0.0
...    ...         ...  ...               ...               ...
763   27.0         0.0  ...               0.0               0.0
835   42.0         0.0  ...               1.0               0.0
1216  40.0         0.0  ...               1.0               0.0
559   19.0         0.0  ...               0.0               0.0
684   33.0         1.0  ...               0.0               1.0

[1137 rows x 11 columns]

**Fitting Multiple Linear Regression Model to Training Set**


> Different types of Regression Algorithm



1. Linear Regression (Univariate or Multivariate)
2. Support Vector Regression
3. Decision Tree Regression
4. Random Forest Regressrion

**Multivariate Linear Regression**

In [29]:
from sklearn.linear_model import LinearRegression

regressor = LinearRegression()
regressor.fit(X_train,y_train)

# Predicting the Test Set Results
y_pred = regressor.predict(X_test)

In [30]:
y_test

charges
578    9724.53000
610    8547.69130
569   45702.02235
1034  12950.07120
198    9644.25250
...           ...
8      6406.41070
998    6551.75010
1217   4058.11610
434    4243.59005
1110  11512.40500

[201 rows x 1 columns]

In [31]:
y_pred

array([[1.10740457e+04],
       [9.52017593e+03],
       [3.81466807e+04],
       [1.61348840e+04],
       [6.93769026e+03],
       [4.15710135e+03],
       [1.73664316e+03],
       [1.42706434e+04],
       [8.94879374e+03],
       [7.44354818e+03],
       [4.70219731e+03],
       [1.02420359e+04],
       [8.70695060e+03],
       [3.87341533e+03],
       [2.79884091e+04],
       [1.07583173e+04],
       [1.12649864e+04],
       [6.21693774e+03],
       [8.18716409e+03],
       [2.73797203e+04],
       [3.37289611e+04],
       [1.42578981e+04],
       [1.18498443e+04],
       [3.22904330e+04],
       [4.17990908e+03],
       [9.32821809e+03],
       [1.30746416e+03],
       [9.81338550e+03],
       [3.83612292e+03],
       [1.03665409e+04],
       [8.95653538e+03],
       [4.01270895e+04],
       [1.55346123e+04],
       [1.37279035e+04],
       [2.50678709e+04],
       [5.30784313e+03],
       [1.25816721e+04],
       [3.07414148e+04],
       [3.35737155e+04],
       [3.55726649e+03],


**Evaluation Metrices**

1. Mean Absolute Error (MAE)
2. Mean Squared Error (MSE)
3. R-Squared Error

In [32]:
from sklearn.metrics import mean_absolute_error

# MAE

mean_absolute_error(y_test, y_pred)

4140.090003563701

In [33]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, y_pred)

34691222.56946

In [34]:
from sklearn.metrics import r2_score

r2_score(y_test, y_pred)

0.7918340172450385

**Feature Scaling**

In [35]:
from sklearn.preprocessing import StandardScaler
X_sc = StandardScaler()
y_sc = StandardScaler()
X_train_new = X_sc.fit_transform(X_train[['age', 'sex_female', 'sex_male',
       'smoker_no', 'smoker_yes', 'bmi','children','region_northeast','region_northwest','region_southeast','region_southwest']])
X_test_new = X_sc.transform(X_test[['age', 'sex_female', 'sex_male',
       'smoker_no', 'smoker_yes', 'bmi','children','region_northeast','region_northwest','region_southeast','region_southwest']])
y_train_new = y_sc.fit_transform(y_train[['charges']])
y_test_new = y_sc.transform(y_test[['charges']])

**Support Vector Regression**

In [36]:
from sklearn.svm import SVR 

regressor = SVR(kernel = 'linear')
regressor.fit(X_train_new,y_train_new)

SVR(kernel='linear')

In [37]:
y_pred = regressor.predict(X_test_new)

In [38]:
r2_score(y_test_new, y_pred)

0.74826138391319

**Decision Tree Regression**

In [39]:

from sklearn.tree import DecisionTreeRegressor


regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X_train_new,y_train_new)

DecisionTreeRegressor(random_state=0)

In [40]:
y_pred = regressor.predict(X_test_new)
r2_score(y_test_new, y_pred)

0.672358754246029

**Random Forest Regression**

In [41]:
from sklearn.ensemble import RandomForestRegressor

regressor = RandomForestRegressor(n_estimators = 10,random_state = 0) # n estiamator is the number of decision trees
regressor.fit(X_train_new,y_train_new) 

RandomForestRegressor(n_estimators=10, random_state=0)

In [42]:
y_pred = regressor.predict(X_test_new)
r2_score(y_test_new, y_pred)

0.8342460442084068

#Result Analysis
> The R-square error for the regression are given below:




1. Linear Regression :0.7918
2. Support Vector Regression:0.74826
3. Decision Tree Regression:0.6723
4. Random Forest Regression:0.8342


> 


Different regression algorithms has been applied on the dataset . For evaluating their performances R-square error evaluation matrix has been applied for all. From the values we can see that, R-square value for random forest regression is more close to 1 which is 0.83 than the others. So that, random forest regression gives better prediction in the dataset. Decision Tree Regression shows less efficient prediction.